src/Microsoft.Health.Fhir.Core/Features/Persistence/BadRequestException.cs

```diff
@@ -18,5 +20,22 @@ public BadRequestException(string errorMessage)
                     OperationOutcomeConstants.IssueType.Invalid,
                     errorMessage));
         }
+
+        /// <summary>
+        /// Constructor creates BadRequestException from collection of error strings.
+        /// </summary>
+        /// <param name="errorMessages">Collection of error strings that are used to initialize OperationOutcomeIssue collection.
+        /// Must be non-null and non-empty</param>
```

Comment:
```C#
            : base(((Func<string>)(() =>
            {
                Debug.Assert(errorMessages != null, "Parameter errorMessages must not be null.");
                Debug.Assert(errorMessages == null || errorMessages.Any(), "Parameter errorMessages must not be empty.");
                return (errorMessages == null || errorMessages.Count < 2) ? errorMessages.First() : "Multiple bad request errors.";
            }))())
            /*: base((errorMessages == null || errorMessages.Count < 2) ? errorMessages.First() : "Multiple bad request errors.")*/
```

Reply:
I almost had this done but then something crossed my mind, Debug.Assert will only show in VS debug or output window. In the code I was debugging yesterday I saw "_output" used and that wrote into the pipelines logs, and I know that for ASP.NET projects using iLogger is proper way. So do we know the right way to output debug messages for on FHIR-server project?

Reply:
Also, just a note...
```C#
: base((errorMessages == null || errorMessages.Count < 2) ? errorMessages.First() : "Multiple bad request errors.")
```
this code does handle null and empty. In both cases First() will be executed and First() will check errorMessages and throw appropriate Exception for both cases (it will check for null as well because First() is static extension).

Reply:
I made expression bit more complex so decide to keep parentheses for clarity.
Decided not to add Debug.Assert, because this would only show in VS debug window. Proper way to log ASP.NET servers is by using ILogger interface https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/ because this can be setup to show messages both when debugging in VS and when deployed on Azure. But I don't see that in the files and should really be a separate item.
Also, any messaging at this point would be redundant because proper exception is thrown by First() for either null or empty.

Reply:
> I almost had this done but then something crossed my mind, Debug.Assert will only show in VS debug or output window. In the code I was debugging yesterday I saw "_output" used in the unit tests and that wrote into the pipelines logs, and I know that for ASP.NET projects using iLogger is proper way. So, do we know the right way to output debug messages for on FHIR-server project?

This is a good question - we use `iLogger` in other parts of the solution, but I'm not sure why we don't use it in the exception classes that inherit from `FhirException`. As discussed, we can post in the fhir server channel to get a bit more info on why `Debug.Assert()` or `EnsureArg.IsNotNull()` is used. Once we decide on a good approach, it seems like we might need a user story to make our null checking across exceptions more uniform.

Reply:
> Also, any messaging at this point would be redundant because proper exception is thrown by First() for either null or empty.

The issue is that throwing an Exception creating an Exception hides the actual cause. This is why the Assert was added as a softer notification in the Exception class constructors. In other places EnsureArgs should be used as it clearly defines which parameter was null.

Reply:
Added Debug.Assert